controllers/edgedeployment_controller.go

```diff
@@ -243,6 +293,29 @@ func (r *EdgeDeploymentReconciler) SetupWithManager(mgr ctrl.Manager) error {
 		Complete(r)
 }
 
+func (r *EdgeDeploymentReconciler) ExecuteConcurrent(funcs []concurrentFunc) []error {
+	nFuncs := len(funcs)
+	if nFuncs == 0 {
+		return nil
+	}
+	returnValues := make([][]error, nFuncs)
+	var wg sync.WaitGroup
+	wg.Add(nFuncs)
+	for i := 0; i < nFuncs; i++ {
```

Comment:
What do you think about about not using slice with functions? We iterate over inputs to create it and later we iterate again to run it. We could only have one loop when we run it.

Reply:
These are different functions and different inputs. I do not see how this can be done without making the code much more complex. Then the loop becomes redundant compared to extra complexity/code.
If you have an idea please suggest.
BTW, does it bother us that much? I mean iterating over the functions? Especially, when we have 100s of EdgeDevices?

Reply:
I would see this something similar to:

```
	inputs := SplitEdgeDevices(edgeDevices, r.Concurrency)
	var s []error

	var wg sync.WaitGroup
	wg.Add(len(inputs))
	
	for _, input := range inputs {
		go func() {
			defer wg.Done()
			err := fn(ctx, name, input)
			if err != nil {
				s = append(s, err)
			}
		}()
	}
	wg.Wait()
	if len(s) != 0 {
		return fmt.Errorf(mergeErrorMessages(s))
	}
	return nil
```

The only issue is to figure out common signature for `fn` if possible to make it more generic. Let me know what you think.

Reply:
I would have done that if it was always the same function.
However, in our case, we have 3 functions and may have more.
So for each function, you need a loop that knows its input and its signature.
I don't think we can work around that. You will always have to create something specific to each function call and will always have to loop over all function calls.
Regarding the channel. I think it is much easier in our case to use a slice.
Our input is known in advance. So we can put it in a collection (e.g. slice).
A channel is an implementation of a queue (for our needs).
It is needed if you don't know when and how many inputs you'll have.
I know that it seems cool to use channels and mocks and a lot of stuff that many people use.
But, it does not mean that it's right.
Your call.

Reply:
I googled 'go generics' and it seems that it's going to be included soon. This is what you're looking for.

Reply:
we can do something like what you suggested if we limit executeConcurrent only for functions that handle []EdgeDevice and nothing else changes from call to call.